File: frontend/src/shared/project.ts
/**
 * `annotation.project.json` - one schema, two outputs: the runtime check and the type.
 *
 * Mirrors `backend/src/annotation_helper/project.py`. Both sides are forgiving in the
 * same way: unknown keys are ignored so a newer file still opens in an older build, and
 * every section has a default so a hand-written file may omit whatever it does not care
 * about. Note the `.default({})` on each object - a zod object whose fields all have
 * defaults is still a *required key* without it.
 */

import { z } from 'zod'

export const PROJECT_FILENAME = 'annotation.project.json'
export const STATE_DIRNAME = '.annotation-helper'
export const FORMAT_VERSION = 1

export const DEFAULT_CLASS_COLORS = [
  '#386938', '#c98a1b', '#3b7f8c', '#8c4a3b', '#5d5b8c',
  '#7a8c3b', '#8c3b6e', '#3b558c', '#8c7a3b', '#4a8c6e'
] as const

export const TASKS = ['detect', 'segment', 'both'] as const
export const PIPELINES = ['detect', 'segment', 'detect_then_segment'] as const

export const projectClassSchema = z.object({
  id: z.number().int().min(0),
  name: z.string().min(1),
  color: z.string().default(DEFAULT_CLASS_COLORS[0])
})

export const projectPathsSchema = z
  .object({
    /** The inbox: images wait here until they are annotated, then move to `images`. */
    input: z.string().default('input'),
    images: z.string().default('images'),
    labels: z.string().default('labels'),
    /** The recycle bin: originals and deletions land here, never `unlink`. */
    recycle: z.string().default('recycle'),
    output: z.string().default('dataset')
  })
  .default({})

export const aiSettingsSchema = z
  .object({
    pipeline: z.enum(PIPELINES).default('detect'),
    detectModel: z.string().default(''),
    segmentModel: z.string().default(''),
    confidence: z.number().min(0).max(1).default(0.25),
    iou: z.number().min(0).max(1).default(0.45),
    maxDetections: z.number().int().min(1).default(300),
    expandRatio: z.number().min(0).max(2).default(0.1),
    simplifyTolerance: z.number().min(0).default(1.5)
  })
  .default({})

export const splitSettingsSchema = z
  .object({
    train: z.number().min(0).default(0.8),
    val: z.number().min(0).default(0.15),
    test: z.number().min(0).default(0.05),
    seed: z.number().int().default(1337)
  })
  .default({})

export const trainingSettingsSchema = z
  .object({
    model: z.string().default('yolo11n.pt'),
    epochs: z.number().int().min(1).default(100),
    imgsz: z.number().int().min(32).default(640),
    batch: z.number().int().min(-1).default(16),
    device: z.string().default(''),
    patience: z.number().int().min(0).default(50),
    runName: z.string().default('train')
  })
  .default({})

export const projectSchema = z.object({
  formatVersion: z.number().int().default(FORMAT_VERSION),
  name: z.string().default('Untitled project'),
  task: z.enum(TASKS).default('detect'),
  classes: z.array(projectClassSchema).min(1).default([{ id: 0, name: 'object', color: DEFAULT_CLASS_COLORS[0] }]),
  paths: projectPathsSchema,
  ai: aiSettingsSchema,
  split: splitSettingsSchema,
  training: trainingSettingsSchema
})

export type ProjectClass = z.infer<typeof projectClassSchema>
export type ProjectPaths = z.infer<typeof projectPathsSchema>
export type AiSettings = z.infer<typeof aiSettingsSchema>
export type SplitSettings = z.infer<typeof splitSettingsSchema>
export type TrainingSettings = z.infer<typeof trainingSettingsSchema>
export type ProjectFile = z.infer<typeof projectSchema>

/** A loaded project: the file, plus where it came from. `root` is never in the file. */
export interface Project extends ProjectFile {
  root: string
}

export function defaultProject(name = 'Untitled project'): ProjectFile {
  return projectSchema.parse({ name })
}

/** Names ordered by id with gaps filled, because YOLO indexes classes by position. */
export function classNames(classes: ProjectClass[]): string[] {
  if (classes.length === 0) return []
  const highest = Math.max(...classes.map((c) => c.id))
  const byId = new Map(classes.map((c) => [c.id, c.name]))
  return Array.from({ length: highest + 1 }, (_, i) => byId.get(i) ?? `class_${i}`)
}

export function classById(classes: ProjectClass[], id: number): ProjectClass | undefined {
  return classes.find((c) => c.id === id)
}

export function classColor(classes: ProjectClass[], id: number): string {
  return classById(classes, id)?.color ?? DEFAULT_CLASS_COLORS[id % DEFAULT_CLASS_COLORS.length]!
}

export function nextClassId(classes: ProjectClass[]): number {
  return classes.length === 0 ? 0 : Math.max(...classes.map((c) => c.id)) + 1
}

File: backend/src/annotation_helper/project.py
"""The project file: `annotation.project.json`.

Design rules, so a project stays usable without this application:

* Every path in the file is *relative to the project folder*. Copy the folder to
  another machine and it still opens.
* The layout is plain ultralytics: `images/`, `labels/`, `classes.txt`, `data.yaml`.
  Any other YOLO tool can read the result; nothing is locked into a private format.
* Application state that is not data - caches, journals, trash - lives in
  `.annotation-helper/` and can be deleted at any time without losing work.
"""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field, replace
from pathlib import Path
from typing import Any

PROJECT_FILENAME = "annotation.project.json"
STATE_DIRNAME = ".annotation-helper"
FORMAT_VERSION = 1

DEFAULT_CLASS_COLORS = [
    "#386938", "#c98a1b", "#3b7f8c", "#8c4a3b", "#5d5b8c",
    "#7a8c3b", "#8c3b6e", "#3b558c", "#8c7a3b", "#4a8c6e",
]


@dataclass(slots=True)
class ProjectClass:
    id: int
    name: str
    color: str = DEFAULT_CLASS_COLORS[0]


@dataclass(slots=True)
class ProjectPaths:
    input: str = "input"  # the inbox; annotating moves an image out of it
    images: str = "images"
    labels: str = "labels"
    recycle: str = "recycle"  # the bin; nothing is unlinked without an explicit empty
    output: str = "dataset"


@dataclass(slots=True)
class AiSettings:
    """Every threshold is a project setting.

    The predecessor hardcoded four different confidence values in four places; this is
    the single home for all of them.
    """

    pipeline: str = "detect"  # detect | segment | detect_then_segment
    detect_model: str = ""
    segment_model: str = ""
    confidence: float = 0.25
    iou: float = 0.45
    max_detections: int = 300
    expand_ratio: float = 0.10  # box growth before the segment crop, pipeline 3 only
    simplify_tolerance: float = 1.5  # pixels; 0 keeps every predicted vertex
    sam_model: str = "mobile_sam.pt"  # click-to-segment; an ultralytics name or a path


@dataclass(slots=True)
class SplitSettings:
    train: float = 0.80
    val: float = 0.15
    test: float = 0.05
    seed: int = 1337


@dataclass(slots=True)
class TrainingSettings:
    model: str = "yolo11n.pt"
    epochs: int = 100
    imgsz: int = 640
    batch: int = 16
    device: str = ""  # "" lets ultralytics choose; "0" or "cpu" to force
    patience: int = 50
    run_name: str = "train"


@dataclass(slots=True)
class Project:
    root: Path
    name: str = "Untitled project"
    task: str = "detect"  # detect | segment | both
    classes: list[ProjectClass] = field(default_factory=lambda: [ProjectClass(0, "object")])
    paths: ProjectPaths = field(default_factory=ProjectPaths)
    ai: AiSettings = field(default_factory=AiSettings)
    split: SplitSettings = field(default_factory=SplitSettings)
    training: TrainingSettings = field(default_factory=TrainingSettings)
    format_version: int = FORMAT_VERSION

    # --- resolved locations -------------------------------------------------

    @property
    def file(self) -> Path:
        return self.root / PROJECT_FILENAME

    @property
    def input_dir(self) -> Path:
        """The inbox. Images wait here until they are annotated."""
        return self.root / self.paths.input

    @property
    def images_dir(self) -> Path:
        return self.root / self.paths.images

    @property
    def labels_dir(self) -> Path:
        return self.root / self.paths.labels

    @property
    def output_dir(self) -> Path:
        return self.root / self.paths.output

    @property
    def state_dir(self) -> Path:
        return self.root / STATE_DIRNAME

    @property
    def recycle_dir(self) -> Path:
        """The bin. A visible folder: what lands here is the user's data, not app state."""
        return self.root / self.paths.recycle

    @property
    def trash_dir(self) -> Path:
        """Kept as the old name for callers; the bin itself moved out of state_dir."""
        return self.recycle_dir

    def class_name(self, class_id: int) -> str:
        for c in self.classes:
            if c.id == class_id:
                return c.name
        return f"class_{class_id}"

    def class_names(self) -> list[str]:
        """Names ordered by id, with gaps filled, because YOLO indexes by position."""
        if not self.classes:
            return []
        highest = max(c.id for c in self.classes)
        return [self.class_name(i) for i in range(highest + 1)]

    # --- persistence --------------------------------------------------------

    def to_dict(self) -> dict[str, Any]:
        return {
            "formatVersion": self.format_version,
            "name": self.name,
            "task": self.task,
            "classes": [asdict(c) for c in self.classes],
            "paths": asdict(self.paths),
            "ai": _camel(asdict(self.ai)),
            "split": asdict(self.split),
            "training": _camel(asdict(self.training)),
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any], root: Path) -> "Project":
        classes = [
            ProjectClass(
                id=int(c.get("id", i)),
                name=str(c.get("name", f"class_{i}")),
                color=str(c.get("color", DEFAULT_CLASS_COLORS[i % len(DEFAULT_CLASS_COLORS)])),
            )
            for i, c in enumerate(data.get("classes", []))
        ] or [ProjectClass(0, "object")]

        return cls(
            root=root,
            name=str(data.get("name", root.name)),
            task=str(data.get("task", "detect")),
            classes=classes,
            paths=_build(ProjectPaths, data.get("paths", {})),
            ai=_build(AiSettings, data.get("ai", {})),
            split=_build(SplitSettings, data.get("split", {})),
            training=_build(TrainingSettings, data.get("training", {})),
            format_version=int(data.get("formatVersion", FORMAT_VERSION)),
        )

    @classmethod
    def load(cls, path: Path) -> "Project":
        """Accepts either the project file or the folder that holds it."""
        path = Path(path)
        file = path / PROJECT_FILENAME if path.is_dir() else path
        if not file.is_file():
            raise FileNotFoundError(f"no {PROJECT_FILENAME} at {path}")
        data = json.loads(file.read_text(encoding="utf-8"))
        return cls.from_dict(data, file.parent)

    def save(self) -> Path:
        """Atomic write, then refresh the two files other tools read."""
        self.root.mkdir(parents=True, exist_ok=True)
        temp = self.file.with_suffix(".json.tmp")
        temp.write_text(
            json.dumps(self.to_dict(), indent=2, ensure_ascii=False) + "\n",
            encoding="utf-8",
            newline="\n",
        )
        temp.replace(self.file)
        self.write_classes_txt()
        return self.file

    def write_classes_txt(self) -> Path:
        """A plain one-name-per-line mirror. Half the YOLO ecosystem expects it."""
        target = self.root / "classes.txt"
        target.write_text("\n".join(self.class_names()) + "\n", encoding="utf-8", newline="\n")
        return target

    def write_data_yaml(self, target: Path | None = None, splits_root: Path | None = None) -> Path:
        """Write the ultralytics dataset descriptor.

        Hand-rolled rather than pulling in PyYAML: the file is a handful of lines and
        the core package stays dependency-free.
        """
        target = target or (self.root / "data.yaml")
        base = splits_root or self.output_dir
        names = self.class_names()
        lines = [
            "# Generated by Annotation Helper. Safe to edit; regenerating overwrites it.",
            f"path: {_yaml_path(base)}",
            "train: train/images",
            "val: val/images",
        ]
        if self.split.test > 0:
            lines.append("test: test/images")
        lines.append(f"nc: {len(names)}")
        lines.append("names:")
        lines.extend(f"  {i}: {name}" for i, name in enumerate(names))
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_text("\n".join(lines) + "\n", encoding="utf-8", newline="\n")
        return target

    def with_changes(self, **changes: Any) -> "Project":
        return replace(self, **changes)


def create_project(root: Path, name: str | None = None, task: str = "detect") -> Project:
    """Lay out a new project folder. Idempotent: re-running never destroys data."""
    root = Path(root)
    project = Project(root=root, name=name or root.name, task=task)
    for directory in (
        project.input_dir,
        project.images_dir,
        project.labels_dir,
        project.recycle_dir,
        project.state_dir,
    ):
        directory.mkdir(parents=True, exist_ok=True)
    (project.state_dir / ".gitignore").write_text("*\n", encoding="utf-8", newline="\n")
    if project.file.exists():
        return Project.load(project.file)
    project.save()
    return project


def find_project(start: Path) -> Path | None:
    """Walk up from `start` looking for a project file, like git looks for `.git`."""
    current = Path(start).resolve()
    for candidate in [current, *current.parents]:
        if (candidate / PROJECT_FILENAME).is_file():
            return candidate
    return None


def _build(cls: type, data: dict[str, Any]):
    """Construct a settings dataclass from camelCase JSON, ignoring unknown keys."""
    fields = set(cls.__dataclass_fields__)  # type: ignore[attr-defined]
    kwargs = {}
    for key, value in (data or {}).items():
        snake = _snake(key)
        if snake in fields:
            kwargs[snake] = value
    return cls(**kwargs)


def _camel(data: dict[str, Any]) -> dict[str, Any]:
    """snake_case -> camelCase, so the JSON reads the same as the TypeScript types."""
    out = {}
    for key, value in data.items():
        head, *rest = key.split("_")
        out[head + "".join(p.title() for p in rest)] = value
    return out


def _snake(key: str) -> str:
    return "".join("_" + c.lower() if c.isupper() else c for c in key)


def _yaml_path(path: Path) -> str:
    """Forward slashes: a Windows backslash in YAML is an escape sequence."""
    return str(path).replace("\\", "/")
